### ISSUE TITLE
[1.12.0-rc] docker service update --image should pull image before stopping services

### ISSUE THREAD

**Issue by BenHall** [docker/docker#24447]
I'm noticing that when I'm using `docker service update --image` it will stop and remove all the containers, pull the image and then restart the containers.

Depending on the size of the application, then can cause the application to be unresponsive for some time. If there is a problem pulling the image (typo), then your application will go down while you're identifying the problem.

Is there a reason not to pull the image first, then apply updates?

**Comment by cpuguy83** [comment]
ping @stevvooe

**Comment by stevvooe** [comment]
@BenHall This is likely due to how the rolling update is occurring. How many replicas are you running for the service?

cc @aluzzardi

**Comment by BenHall** [comment]
@stevvooe When using rolling updates the first one takes the hit so it's not as much of a problem. This happens no matter how many replicas are running.

**Comment by stevvooe** [comment]
> This happens no matter how many replicas are running.

I ask because the affect will be worse with fewer replicas.

It sounds like we may be routing traffic to this task, even though it may not actually be running yet.

cc @mrjana

**Comment by cpuguy83** [comment]
Ok, so I just tested this by scaling a service to 5, then updating the image to a large image that I didn't have already, it takes down all services even with `--update-parallelism=1`.
To me this is pretty bad.

**Comment by thaJeztah** [comment]
@stevvooe who would be the best person to assign to this?

**Comment by stevvooe** [comment]
@thaJeztah Confirmed this is completely broken. Looking into it now.

**Comment by aluzzardi** [comment]
ping @aaronlehmann

**Comment by stevvooe** [comment]
Did one more test with `--update-parallelism` set to 1.

There are few compounding issues:

1. For small images, `--update-parallelism 1` will do the trick. We are going to make this the default. Adding `--update-delay` may also help here. (I'll submit a PR for this shortly)
2. There seems to be a race in image pull and container create. I am seeing `No such image: ubuntu:15.10` in certain condition. (needs to be investigated)
3. Rolling update is ignoring errors proceeding. (https://github.com/docker/docker/pull/24943 to addresses this)
4. We probably need to coordinate such that tasks are staged to `READY` before killing old tasks (won't fix in 1.12 but will have something more sophisticated for 1.13).

cc @aluzzardi @aaronlehmann

**Comment by stevvooe** [comment]
I've created a gist with pull race condition https://gist.github.com/stevvooe/5596fd68374b3f03f3490cc06289b14c.

We may want a second issue for that.

cc @aaronlehmann

**Comment by aaronlehmann** [comment]
Point 1 will be fixed by #24952
Point 2 will be fixed by https://github.com/docker/swarmkit/pull/1212 once swarmkit is revendored.
Point 3 should be fixed by #24943
Point 4 will be addressed in the 1.13 timeframe.

**Comment by asantos82** [comment]
Hi,

I am using this to command to "update" my stack. ```docker stack deploy --with-registry-auth --compose-file docker-stack-prd-w.yml odoo```
From the previous deploy, to the new one what changes is the image tag in the docker-stack-prd-w.yml, so I want to update my service from tag1 to tag2.

The docker-stack-prd-w.yml file has the following config
```
    deploy:
      replicas: 1
      update_config:
        parallelism: 1
        delay: 10s
      restart_policy:
        condition: on-failure
        delay: 10s
        max_attempts: 30
        window: 120s
```
What I see is that the service is stopped, then the image is downloaded to the node, and then the service is started with the new image.
This image is big, aroung 1.1GB, so the service will be down for a lot of time.

Is there any why, to perform the actions in the following order? Download the image to the node, stop the service, and then start the service with the new image?

```
$ docker version 
Client:
 Version:      17.03.0-ce
 API version:  1.26
 Go version:   go1.7.5
 Git commit:   60ccb22
 Built:        Thu Feb 23 10:57:47 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.03.0-ce
 API version:  1.26 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   60ccb22
 Built:        Thu Feb 23 10:57:47 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Comment by aaronlehmann** [comment]
Essentially it's because there is 1 replica specified, so the reconciliation loop tries to keep only one task active at any moment. This means the old task is shut down at the same time that the new one is created, and the new one needs to pull before it is able to run.

In 17.05 we're adding a flag that allows the old task to stay running until the new one has finished starting up: see #30261. Note that this means there will be a brief interval where two replicas are running instead of one. Also, support hasn't been added to compose files yet (see #32586).

In the future we might add an option or change the default so that the image gets pulled before any tasks get shut down.

**Comment by thaJeztah** [comment]
> In the future we might add an option or change the default so that the image gets pulled before any tasks get shut down.

I think changing the default would make sense; at least it guarantees we were able to pull the image before the task is shutdown

**Comment by asantos82** [comment]
@thaJeztah Yes, pulling the image to the node, prior to shutting down services, would reduce significantly the "downtime"

**Comment by Dzhuneyt** [comment]
@thaJeztah @stevvooe any movement on this? Or a potential ETA.

**Comment by JulyIghor** [comment]
Any solution years later?

**Comment by JulyIghor** [comment]
Any solution years later?

**Comment by thaJeztah** [comment]
@dperny PTAL

**Comment by JulyIghor** [comment]
this issue brings additional problem that if service image is based on some other big image, it is being deleted even if not changed, so first it have to pull, than removing old image will not cause deleting parent image and save lots of time
